packages/itwin/tree-widget/src/components/trees/stateless/models-tree/ModelsTree.tsx

```diff
@@ -210,3 +216,42 @@ function useCachedVisibility(activeView: Viewport, hierarchyConfig: ModelsTreeHi
     visibilityHandlerFactory,
   };
 }
+
+function ClassGroupingIcon() {
+  return (
+    <Icon>
+      <svg id="Calque_1" data-name="Calque 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16">
+        <path d="M8.00933,0,0,3.97672V11.986L8.00933,16,16,11.93V3.97651ZM1.66173,11.27642c-.26155.03734-.59754-.26154-.76553-.69085-.168-.41066-.09334-.784.168-.82152.26154-.03734.59754.26154.76553.67219C1.99772,10.86577,1.92306,11.23909,1.66173,11.27642Zm0-3.32319c-.26155.03733-.59754-.28-.76553-.69086-.168-.42932-.09334-.80285.168-.84.26133-.03733.59754.28.76532.69086C1.99772,7.54236,1.92306,7.89723,1.66173,7.95323Zm4.31276,5.52621a.18186.18186,0,0,1-.16821-.01866L3.41657,12.15394a.94275.94275,0,0,1-.29887-.80285c.03754-.33621.22421-.52265.41108-.41066L5.9185,12.24727a.88656.88656,0,0,1,.28.80285A.5057.5057,0,0,1,5.97449,13.47944Zm0-3.37919a.18184.18184,0,0,1-.16821-.01867L3.41657,8.77475a.943.943,0,0,1-.29887-.80286c.03754-.3362.22421-.52286.41108-.42953L5.9185,8.86786a.83112.83112,0,0,1,.28.78419A.51684.51684,0,0,1,5.97449,10.10025Z" />
+      </svg>
+    </Icon>
+  );
+}
+
+async function collectTargetKeys(loadFocusedInstancesKeys: () => AsyncIterableIterator<InstanceKey | GroupingHierarchyNode>) {
+  const targetKeys: Array<InstanceKey | ElementsGroupInfo> = [];
+  for await (const key of loadFocusedInstancesKeys()) {
+    if ("id" in key) {
+      targetKeys.push(key);
+      continue;
+    }
+
+    if (!HierarchyNodeKey.isClassGrouping(key.key)) {
+      targetKeys.push(...key.groupedInstanceKeys);
+      continue;
+    }
+
+    if (!key.nonGroupingAncestor || !HierarchyNodeKey.isInstances(key.nonGroupingAncestor.key)) {
+      continue;
+    }
+
+    const parentKey = key.nonGroupingAncestor.key.instanceKeys[0];
+    const type = key.nonGroupingAncestor.extendedData?.isCategory ? "category" : "element";
+    const targetInfo = targetKeys.find((target): target is ElementsGroupInfo => !("id" in target) && target.parentKey.id === parentKey.id);
+    if (targetInfo) {
+      targetInfo.classes.push(key.key.className);
```

Comment:
shouldn't we worry about duplicate class names here?

Reply:
I assume there should not be multiple grouping nodes for same class under same non-grouping parent